File: WebApps/AngularClient/Services/CategoryService.cs
using ApiClient.Catalog.Category;
using ApiClient.Catalog.Category.Models;
using ApiClient.Catalog.Models.Catalog.Category;
using ApiClient.Common;
using Platform.ApiBuilder;

namespace AngularClient.Services;

public interface ICategoryService
{
    Task<List<CategorySummary>?> GetCategoriesAsync(CancellationToken cancellationToken = default);
    Task<CategoryDetail?> GetCategoryByIdAsync(string id, CancellationToken cancellationToken = default);
    Task<CategoryDetail?> GetCategoryByNameAsync(string name, CancellationToken cancellationToken = default);
    Task<CategoryDetail?> CreateCategoryAsync(CreateCategoryRequestBody requestBody, CancellationToken cancellationToken = default);
    Task<CategoryDetail?> UpdateCategoryAsync(UpdateCategoryRequestBody requestBody, CancellationToken cancellationToken = default);
    Task<ApiStatusResult> DeleteCategoryAsync(string id, CancellationToken cancellationToken = default);
}
public class CategoryService : ICategoryService
{
    private readonly ICategoryApiClient _categoryApiClient;
    public CategoryService(ICategoryApiClient categoryApiClient)
    {
        _categoryApiClient = categoryApiClient;
    }
    public async Task<List<CategorySummary>?> GetCategoriesAsync(CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.GetCategoriesAsync(cancellationToken);

        if (result.IsSuccessStatusCode && result.Result is not null)
        {
            return result.Result.ToList();
        }

        return null;
    }

    public async Task<CategoryDetail?> GetCategoryByIdAsync(string id, CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.GetCategoryByIdAsync(id, cancellationToken);

        if (result.IsSuccessStatusCode && result.Result is not null)
        {
            return result.Result;
        }

        return null;
    }

    public async Task<CategoryDetail?> GetCategoryByNameAsync(string name, CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.GetCategoryByNameAsync(name, cancellationToken);

        if (result.IsSuccessStatusCode && result.Result is not null)
        {
            return result.Result;
        }

        return null;
    }

    public async Task<CategoryDetail?> CreateCategoryAsync(CreateCategoryRequestBody requestBody, CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.CreateCategoryAsync(requestBody, cancellationToken);

        if (result.IsSuccessStatusCode && result.Result is not null)
        {
            return result.Result;
        }

        return null;
    }

    public async Task<CategoryDetail?> UpdateCategoryAsync(UpdateCategoryRequestBody requestBody, CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.UpdateCategoryAsync(requestBody, cancellationToken);

        if (result.IsSuccessStatusCode && result.Result is not null)
        {
            return result.Result;
        }

        return null;
    }

    public async Task<ApiStatusResult> DeleteCategoryAsync(string id, CancellationToken cancellationToken)
    {
        var result = await _categoryApiClient.DeleteCategoryAsync(id, cancellationToken);

        return result;
    }
}

File: WebApps/AngularClient/ClientApp/src/app/core/service/catalog/category.service.ts
import { Injectable } from "@angular/core";
import { ApiService } from "../api.service";
import { CategorySummary } from "../../models/catalog/category-models/category-summary.model";
import { environment } from "src/environments/environment";
import { HttpClient, HttpParams } from "@angular/common/http";
import { CategoryDetail } from "../../models/catalog/category-models/category-detail.model";
import { lastValueFrom } from "rxjs";

@Injectable()
export class CategoryService {
  apiName: string = "Category";

  constructor(public apiService: ApiService, private httpClient: HttpClient) {

  }

  //Todo: Example for HttpClient
  async getCategoriesAsync(): Promise<CategorySummary[]> {
    let observableResult = this.httpClient.get(`${environment.baseApiUrl}/${this.apiName}/GetCategories`);

    return await lastValueFrom(observableResult) as CategorySummary[];
  }

  async getCategoryByIdAsync(id: string): Promise<CategoryDetail> {
    return await this.apiService.getAsync(`${environment.baseApiUrl}/${this.apiName}/GetCategoryById/${id}`, new HttpParams());
  }

  async getCategoryByNameAsync(name: string): Promise<CategoryDetail> {
    return await this.apiService.getAsync(`${environment.baseApiUrl}/${this.apiName}/GetCategoryByName/${name}`, new HttpParams());
  }

  async createCategoryAsync(categoryDetail: any): Promise<CategoryDetail> {
    return await this.apiService.postAsync(`${environment.baseApiUrl}/${this.apiName}/CreateCategory`, categoryDetail, new HttpParams());
  }

  async updateCategoryAsync(categoryDetail: any): Promise<CategoryDetail> {
    return await this.apiService.putAsync(`${environment.baseApiUrl}/${this.apiName}/UpdateCategory`, categoryDetail, new HttpParams());
  }

  async deleteCategoryAsync(id: string): Promise<boolean> {
    return await this.apiService.deleteAsync(`${environment.baseApiUrl}/${this.apiName}/DeleteCategory/${id}`, new HttpParams());
  }
}
